src/goto-programs/loop_ids.h

```diff
@@ -31,6 +31,15 @@ struct loop_idt
   {
   }
 
+  loop_idt() : function_id(""), loop_number(-1)
+  {
+  }
+
+  loop_idt(const loop_idt &other)
+    : function_id(other.function_id), loop_number(other.loop_number)
+  {
+  }
+
```

Comment:
Which piece of code needs these? At least the first one looks a bit dangerous and I wonder whether we might instead need to go for a bigger change where `loop_number` is made `optionalt<unsigned int>`.

Reply:
When computing cause loop, we want a value to indicate that no cause loop found, which means that the violation is nothing about loop invariants. Therefore, synthesizing loop invariant doesn't help in such case.
Also, now `loop_number` is a member variable for every instruction. However, it make no sense to have a loop number for an instruction not in a loop. So I agree that changing `loop_number` from `unsigned int` to `optionalt<unsigned int>` is a good idea.
I will make the change for `loop_idt` in this PR. And then change other `loop_number` with another PR. What do you think?

Reply:
Do return an `optionalt<loop_idt>` when looking for a cause.
Please don't change `goto_programt::instructiont`. It's deliberate that the `loop_number` in the instruction is undefined unless the instruction forms a loop.